src/core/smack.c

```diff
@@ -0,0 +1,953 @@
+{
+    huffnode8 *node = &tree->nodes[tree->size++];
+    if (read_bit(bs)) {
+        node->is_leaf = 0;
+        node->b[0] = build_tree8_nodes(bs, tree, value, bits + 1);
+        node->b[1] = build_tree8_nodes(bs, tree, value | (1 << bits), bits + 1);
+    } else {
+        node->is_leaf = 1;
+        node->value = read_byte(bs);
+    }
+    return node;
+}
+
+static hufftree8 *create_tree8(bitstream *bs)
+{
+    if (read_bit(bs)) {
+        hufftree8 *tree = (hufftree8 *) smk_malloc(sizeof(hufftree8));
+        if (!tree) {
+            log_error("SMK: no memory for 8-bit tree", 0, 0);
+            return NULL;
+        }
+        build_tree8_nodes(bs, tree, 0, 0);
+        if (read_bit(bs) != 0) {
+            log_error("SMK: 8-bit tree not closed", 0, 0);
+            free(tree);
+            return NULL;
+        }
+        return tree;
+    } else {
+        log_info("SMK: WARN: no 8-bit tree found", 0, 0);
+        return NULL;
+    }
+}
+
+static void free_tree8(hufftree8 *tree)
+{
+    free(tree);
+}
+
+static uint8_t lookup_tree8(bitstream *bs, hufftree8 *tree)
+{
+    huffnode8 *node = &tree->nodes[0];
+    while (!node->is_leaf) {
+        node = node->b[read_bit(bs)];
+    }
+    return node->value;
+}
+
+// 16-bit huffman tree functions
+
+static void free_node16(huffnode16 *node)
+{
+    if (!node) {
+        return;
+    }
+    if (!node->is_leaf) {
+        free_node16(node->b[0]);
+        free_node16(node->b[1]);
+    }
+    free(node);
+}
+
+static void free_tree16(hufftree16 *tree)
+{
+    if (!tree) {
+        return;
+    }
+    for (int i = 0; i < 3; i++) {
+        if (!tree->escape_nodes[i]->is_leaf) {
+            // Free manually allocated node because it's not in the tree
+            free(tree->escape_nodes[i]);
+        }
+    }
+    free_node16(tree->root);
+    free_tree8(tree->low);
+    free_tree8(tree->high);
+    free(tree);
+}
+
+static huffnode16 *build_tree16_nodes(bitstream *bs, hufftree16 *tree, int value, int bits)
+{
+    huffnode16 *node = (huffnode16 *) smk_malloc(sizeof(huffnode16));
+    if (!node) {
+        log_error("SMK: no memory for 16-bit tree node", 0, 0);
+        return NULL;
+    }
+    if (read_bit(bs)) {
+        node->is_leaf = 0;
+        node->b[0] = build_tree16_nodes(bs, tree, value, bits + 1);
+        if (!node->b[0]) {
+            free(node);
+            return NULL;
+        }
+        node->b[1] = build_tree16_nodes(bs, tree, value | (1 << bits), bits + 1);
+        if (!node->b[1]) {
+            free_node16(node->b[0]);
+            free(node);
+            return NULL;
+        }
+    } else {
+        node->is_leaf = 1;
+        uint8_t lo_val = lookup_tree8(bs, tree->low);
+        uint8_t hi_val = lookup_tree8(bs, tree->high);
+        uint16_t leaf_value = lo_val | (hi_val << 8);
+        node->value = leaf_value;
+
+        for (int i = 0; i < 3; i++) {
+            if (leaf_value == tree->escape_codes[i]) {
+                tree->escape_nodes[i] = node;
```

Comment:
This may not make a difference for our videos, but shouldn't `node->value` be set to `0` here? Or do you do it somewhere else?

As per the [documentation](https://wiki.multimedia.cx/index.php/Smacker#Optimized_Compression): `5a: When you read Leaf and it's value matches one of the markers obtained on step 4, remember the leaf for this marker and set leaf's value to zero.`

Reply:
Exactly :)